keps/sig-node/4589-image-pull-source-support/README.md

```diff
@@ -0,0 +1,972 @@
+
+## Motivation
+
+As the Kubernetes is evolving in the cloud sector and users are willing to have their services offered in a disconnected/private environments, it is necessary to mirror the respective container images internally. Currently there is no easier way to determine the origin of the pulled image when image mirrors are configured other than going through the runtime logs.
+
+### Goals
+
+* Container Status section should show the actual pull source of the image
+* Inspecting Container Image using cri-tools should show the actual pull source of the image
+
+### Non-Goals
+
+* Not concerned about the image field in the container spec which may or may not be the same as the one being pulled by runtime.
+* Not concerned about the mirrors which are configured only for digest pulls.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This KEP aims to introduce a new field in the Container Status as well as in the Container Image resections, inorder to provide the exact registry details of a image pulled by the runtime irrespective of the mirroring configuration. 
+* When no mirrors are configured, the registry details in the new field should be same as the one mentioned in the container spec section
+* When mirrors are configured, the image source details in the new field should be the one runtime used to pull that image.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+* When the imagePullPolicy is set to "Always" and when there are more than one 
+  mirrors configured for that image, the kubelet queries the container image registry 
+  to resolve the name to an image digest. If the kubelet has a container image with that exact 
+  digest cached locally, the kubelet uses its cached image and in this case, it is possible that 
+  the image digest might have been pulled from different source but the imagePullSource 
+  would still be pointing to an existing pulled source.
+* If the image is pulled outside of Kubernetes ecosystem, the imageSource field won't be set 
+  in the Image status. Hence its not possible to find the pull source of that image.
```

Comment:
any image may have a plurality of sources.. one per container per pod potentially .. 

each pod can have a runtime handler and that can have a different snapshotter pulling the image at a different point in time..

pod b does not have access to the private image pulled for pod A, the image may need to be pulled again... mirror destinations can change... that would result in at least two pulls of the same image where the second was just to verify the manifest... but potentially the first is the one used and has the pull location stored.. if they are both using the same snapshotter...

Reply:
Valid point. If the same images are private for each pods and from different source, it might cause the inconsistency result in the pull source. This now leads me to consider the manifest pulls also. Does it mean, considering the manifest pull as the actual source would make more sense than blobs source? This scenario excludes the fact that mirrors are configured for manifest pull only.